### PR TITLE
ISSUE-786 Cassandra implementation of JmapSettingsRepository

### PR DESCRIPTION
**Pull request by thanhbv200585**

### CODE REVIEW

File: tmail-backend/jmap/extensions-cassandra/pom.xml [hunk 1]
```diff
@@ -54,6 +54,11 @@
             <artifactId>testcontainers</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions-cassandra/pom.xml:58]: -> 
`<groupId>${james.groupId}</groupId>`

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    SMono.fromPublisher(executor.executeSingleRow(selectOne.bind()
+      .set(USER, username.asString(), TypeCodecs.TEXT))
+      .map(toJmapSettings))
+
+  def selectState(username: Username): SMono[UuidState] =
+    SMono.fromPublisher(executor.executeSingleRow(selectOne.bind()
+      .set(USER, username.asString(), TypeCodecs.TEXT))
+      .map(toState))
+
+  def updateSetting(username: Username, newState: Option[UuidState], newSettings: Option[Map[JmapSettingsKey, JmapSettingsValue]]): SMono[Void] = {
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:93]: Option[UuidState] -> UuidState 
when we update setting, we always have a new state

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+  private val selectAll: PreparedStatement = session.prepare(selectFrom(TABLE_NAME)
+    .all()
+    .whereColumn(USER).isEqualTo(bindMarker(USER))
+    .build())
+
+  private val selectOne: PreparedStatement = session.prepare(selectFrom(TABLE_NAME)
+    .all()
+    .whereColumn(USER).isEqualTo(bindMarker(USER))
+    .build())
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:61]: What is the difference between selectAll and selectOne ?

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    .value(USER, bindMarker(USER))
+    .value(STATE, bindMarker(STATE))
+    .value(SETTINGS, bindMarker(SETTINGS))
+    .build())
+
+  private val updateStatement: PreparedStatement = session.prepare(update(TABLE_NAME)
+    .setColumn(STATE, bindMarker(STATE))
+    .setColumn(SETTINGS, bindMarker(SETTINGS))
+    .where(column(USER).isEqualTo(bindMarker(USER)))
+    .build())
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:51]: In cassandra insert == update

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  def selectAll(username: Username): SFlux[JmapSettings] =
+    SFlux.fromPublisher(executor.executeRows(selectAll.bind()
+      .set(USER, username.toString, TypeCodecs.TEXT))
+      .map(toJmapSettings))
+
+  def selectOne(username: Username): SMono[JmapSettings] =
+    SMono.fromPublisher(executor.executeSingleRow(selectOne.bind()
+      .set(USER, username.asString(), TypeCodecs.TEXT))
+      .map(toJmapSettings))
+
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:87]: What is the difference?

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala (hunk 2 above)
Comment (vttranlina) [reply tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:93]: Same with `newSettings: Option[Map[JmapSettingsKey, JmapSettingsValue]]`

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsRepository.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  override def updatePartial(username: Username, settingsPatch: JmapSettingsPatch): Publisher[SettingsStateUpdate] = {
+    Preconditions.checkArgument(!settingsPatch.isEmpty, "Cannot update when upsert and remove is empty".asInstanceOf[Object])
+    Preconditions.checkArgument(!settingsPatch.isConflict, "Cannot update and remove the same setting key".asInstanceOf[Object])
+    val newState: UuidState = JmapSettingsStateFactory.generateState()
+    val oldState: UuidState = dao.selectState(username).block()
+
+    SMono.fromCallable(() => {
+      dao.updateSetting(username, Some(newState), Some(settingsPatch.toUpsert.settings))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsRepository.scala:31]: How about values to delete?

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  def deleteOne(username: Username): SMono[Void] =
+    SMono.fromPublisher(executor.executeVoid(deleteOne.bind()
+      .set(USER, username.asString(), TypeCodecs.TEXT)))
+
+  private def toJmapSettings(row: Row): JmapSettings = {
+    JmapSettings(settings = toSettings(row), state = UuidState(row.get(STATE, TypeCodecs.UUID)))
+  }
+
+  private def toSettings(row: Row): Map[JmapSettingsKey, JmapSettingsValue] =
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:118]: can you check it again?
It is "red" on my IDE
image

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsRepository.scala [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    SMono.fromCallable(() => {
+      dao.updateSetting(username, Some(newState), Some(settings.settings))
+    }).`then`(SMono.just(SettingsStateUpdate(oldState, newState)))
+  }
+
+  override def updatePartial(username: Username, settingsPatch: JmapSettingsPatch): Publisher[SettingsStateUpdate] = {
+    Preconditions.checkArgument(!settingsPatch.isEmpty, "Cannot update when upsert and remove is empty".asInstanceOf[Object])
+    Preconditions.checkArgument(!settingsPatch.isConflict, "Cannot update and remove the same setting key".asInstanceOf[Object])
+    val newState: UuidState = JmapSettingsStateFactory.generateState()
+    val oldState: UuidState = dao.selectState(username).block()
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsRepository.scala:28]: can we get the current settings from db, and compare with the settings from the request, if it change then we will update to db

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+  private val deleteOne: PreparedStatement = session.prepare(deleteFrom(TABLE_NAME)
+    .whereColumn(USER).isEqualTo(bindMarker(USER))
+    .build())
+
+  def insert(username: Username, jmapSettings: JmapSettings): SMono[JmapSettings] = {
+    val settings: Map[String, String] = jmapSettings.settings.map(item => (item._1.value.value, item._2.value))
+    val insertSetting = insert.bind()
+      .set(USER, username.asString(), TypeCodecs.TEXT)
+      .set(STATE, jmapSettings.state.value.toString, TypeCodecs.TEXT)
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:71]: jmapSettings.state.value.toString -> jmapSettings.state.serialize

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsRepository.scala [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+  override def reset(username: Username, settings: JmapSettingsUpsertRequest): Publisher[SettingsStateUpdate] = {
+    val newState: UuidState = JmapSettingsStateFactory.generateState()
+    val oldState: UuidState = dao.selectState(username).block()
+    SMono.fromCallable(() => {
+      dao.updateSetting(username, Some(newState), Some(settings.settings))
+    }).`then`(SMono.just(SettingsStateUpdate(oldState, newState)))
+  }
+
+  override def updatePartial(username: Username, settingsPatch: JmapSettingsPatch): Publisher[SettingsStateUpdate] = {
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsRepository.scala:24]: I have serious doubts on the commutativity of your "updatePartial" method" and strongly believes the contract previously written to be insufficient. (set do not commutes)

Do we have a test:

```
update partial - insert (k1 -> v1)
update partial - insert (k2 -> v2)

Expected result (k1 -> v1 + k2 -> v2)
```

Also concurrent testing can be helpful to check for a valid CRDT (using ConcurrentTestRunner)

```
With 10 thread tx do 10 options oy
foreach do: partial update - insert (kx-y -> vx-y)
Expected result: settings with 100 entries, right values
```

(Cassandra have a specific semantic for maps partial updates that MUST be used here)

Comment (vttranlina) [reply tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsRepository.scala:24]: > Do we have a test:
> 
> ```
> update partial - insert (k1 -> v1)
> update partial - insert (k2 -> v2)
> 
> Expected result (k1 -> v1 + k2 -> v2)
> ```
> 

It alreadys. 
`JmapSettingsRepositoryContract#updatePartialShouldAppendNewSettingEntryInPatch`

Comment (quantranhong1999) [reply tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsRepository.scala:24]: I think Benoit' idea here is more testing partial update concurrently on different setting entries.

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    SMono.fromPublisher(executor.executeVoid(updateStatementBuilder.build()))
+  }
+
+  def deleteOne(username: Username): SMono[Void] =
+    SMono.fromPublisher(executor.executeVoid(deleteOne.bind()
+      .set(USER, username.asString(), TypeCodecs.TEXT)))
+
+  private def toJmapSettings(row: Row): JmapSettings = {
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:114]: `{}` is uneeded

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+import com.datastax.oss.driver.api.querybuilder.QueryBuilder.{bindMarker, deleteFrom, insertInto, selectFrom, update}
+import com.datastax.oss.driver.api.querybuilder.relation.Relation.column
+import com.google.common.collect.ImmutableMap
+import com.linagora.tmail.james.jmap.settings.JmapSettings.{JmapSettingsKey, JmapSettingsValue}
+import javax.inject.Inject
+import org.apache.james.backends.cassandra.components.CassandraModule
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor
+import org.apache.james.core.Username
+import org.apache.james.jmap.core.UuidState
+import reactor.core.scala.publisher.{SFlux, SMono}
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:17]: Hmmm I think you might need some import configuration on your IDE, I'm not convinced this is gonna pass the checkstyle

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala (hunk 5 above)
Comment (thanhbv200585) [reply tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:87]: SelectAll uses executeRows but SelectOne uses executeSingleRow. But I think selectAll is unnecessary

**Comment by Arsnael** [comment]
```
[ERROR] /home/jenkins/build/workspace/Tmail_build_PR-799/tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:72: type mismatch;
 found   : com.datastax.oss.driver.api.core.type.codec.TypeCodec[com.google.common.collect.ImmutableMap[String,String]]
 required: com.datastax.oss.driver.api.core.type.codec.TypeCodec[Object]
Note: com.google.common.collect.ImmutableMap[String,String] <: Object, but Java-defined trait TypeCodec is invariant in type JavaTypeT.
You may wish to investigate a wildcard type such as `_ <: Object`. (SLS 3.2.10)
[ERROR] /home/jenkins/build/workspace/Tmail_build_PR-799/tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:104: type mismatch;
 found   : com.datastax.oss.driver.api.core.type.codec.TypeCodec[com.google.common.collect.ImmutableMap[String,String]]
 required: com.datastax.oss.driver.api.core.type.codec.TypeCodec[Object]
Note: com.google.common.collect.ImmutableMap[String,String] <: Object, but Java-defined trait TypeCodec is invariant in type JavaTypeT.
You may wish to investigate a wildcard type such as `_ <: Object`. (SLS 3.2.10)
[ERROR] /home/jenkins/build/workspace/Tmail_build_PR-799/tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/settings/CassandraJmapSettingsDAO.scala:120: value map is not a member of com.google.common.collect.ImmutableMap[String,String]
possible cause: maybe a semicolon is missing before `value map`?
[ERROR] three errors found
```

**Review by chibenwa — APPROVED** [review]

**Comment by Arsnael** [comment]
```
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile (scala-compile-first) on project jmap-extensions-cassandra: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile failed: Compilation failed: InterfaceCompileFailed -> [Help 1]
```
some scala compilation issues :)

**Comment by vttranlina** [comment]
TODO rebase 
and don't forget we have a new api: delete
https://github.com/linagora/tmail-backend/pull/810

**Review by Arsnael — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/JmapSettingsGet.scala [hunk 13]
```diff
@@ -11,25 +11,25 @@ case class JmapSettingsGet(accountId: AccountId,
                            ids: Option[Set[JmapSettingsId]] = None) extends WithAccountId
 case class JmapSettingsResponse(accountId: AccountId,
                                 state: UuidState,
-                                list: Seq[JmapSettingsObject],
+                                list: Seq[JmapSettingsEntry],
                                 notFound: Seq[JmapSettingsId])
 
-object JmapSettingsObject {
+object JmapSettingsEntry {
```
Comment (quantranhong1999) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/JmapSettingsGet.scala:17]: `JmapSettingsEntry` -> `JmapSettingsObject` in the `ISSUE-788 Setting to enable Push for firebase (#794)` commit. To better distinguish with `trait JmapSettingEntry `.

Why change this back?

Comment (thanhbv200585) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/JmapSettingsGet.scala:17]: Bacause I see the latest commit use `JmapSettingsEntry`. I will change it into `JmapSettingsObject`.

**Review by quantranhong1999 — APPROVED** [review]

**Review by vttranlina — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by vttranlina** [comment]
Can you re-squash? I see you squash ISSUE-788 and 786 to single commit, 
It shouldn't, It is two seprate commit

**Pull request merged**